### DISCUSSION TITLE
[Bug]: Configuration with TailwindCss

### DISCUSSION THREAD

**Discussion by tibo-glamarche** [storybookjs/storybook#26323] [Help] [answered]
### Describe the bug

I have followed the procedure in the documentation and the proposed solution of other similar issues but it still doesn't work.Here is what I have tried:
1. Add my index.css file in preview.tsx
2.  I tried importing "tailwindcss/tailwind.css", but that doesn't work.
3.  I messed around with tailwind configs. 

In my app, I import the same index.css and it works.

### System

```bash
Storybook Environment Info:

  System:
    OS: Windows 11 10.0.22621
    CPU: (16) x64 AMD Ryzen 7 PRO 6850U with Radeon Graphics
  Binaries:
    Node: 18.16.1 - C:\Program Files\nodejs\node.EXE
    npm: 9.8.0 - C:\Program Files\nodejs\npm.CMD <----- active
  Browsers:
    Edge: Chromium (120.0.2210.61)
  npmPackages:
    @storybook/addon-essentials: ^7.4.5 => 7.4.5
    @storybook/addon-interactions: ^7.4.5 => 7.4.5
    @storybook/addon-links: ^7.4.5 => 7.4.5
    @storybook/blocks: ^7.4.5 => 7.4.5
    @storybook/react: ^7.4.5 => 7.4.5
    @storybook/react-vite: ^7.4.5 => 7.4.5
    @storybook/testing-library: ^0.2.1 => 0.2.1
    storybook: ^7.4.5 => 7.4.5
```

### Additional context

I am using React + vite
Tailwind v3.3.3

**Comment by fxmkdev** [comment]
I also have issues getting Tailwind CSS to work with a vite-based Storybook. In my case, it is a new Remix app with its built-in Tailwind CSS support. 

I did the following:
1. Set up Storybook, choosing Vite as the bundler
2. Work around #25256 by adding the following to `.storybook/main.ts`:
  ```ts
  async viteFinal(config) {
    // Merge custom configuration into the default config
    return mergeConfig(config, {
      assetsInclude: ["/sb-preview/runtime.js"], // Bug workaround, see https://github.com/storybookjs/storybook/issues/25256
    });
  },
  ```
3. Add `import "../app/tailwind.css";` to `.storybook/preview.ts`. This is the file with the `@tailwind` directives.

When I inspect the preview, I see that the CSS file is present, but the `@tailwind` directives have not been replaced:

<img width="922" alt="image" src="https://github.com/storybookjs/storybook/assets/11920943/68b4a085-7e6e-48a1-845f-cfa4d61a8968">

Is this an issue with PostCSS? To my understanding, PostCSS doesn't need to be configured in vite-based storybooks.

**Comment by fxmkdev** [comment]
Ok, for Remix, I figured it out: while Remix automatically knows how to use Tailwind CSS, PostCSS needs to be configured for Storybook so that it knows that we want to use Tailwind CSS:

1. Add the `autoprefixer` PostCSS plugin (Remix doesn't come with it): `npm i -D autoprefixer`
2. Add a `postcss.config.cjs` file to the project as per the [Tailwind CSS documentation](https://tailwindcss.com/docs/using-with-preprocessors#vendor-prefixes):
  ```js
  // Explicit PostCSS config file to support Tailwind CSS in Storybook
  module.exports = {
    plugins: {
      tailwindcss: {},
      autoprefixer: {},
    },
  };
  ```
3. Run Storybook

The Tailwind CSS stylings are applied now.

@tibo-glamarche Not sure if this is the same issue in your case. Make sure your project has a PostCSS config file.

**Comment by tibo-glamarche** [comment]
I already have a postcss config file with the same configuration as you, but it doesn't work. Thanks anyway!

**Comment by shilman** [comment]
@Integrayshaun is this something you can help with?

**Comment by alejomejia** [comment]
Same problem here using NextJS

**Comment by ShaunEvening** [comment]
@tibo-glamarche @alejomejia 

can you share a reproduction of your problem so that I can dig in?

**Comment by rafalwawrzyk** [comment]
any progress with this? same story with TurboRepo and Next.js, tailwind classes not applied even with direct css file with tailwind config. Only thing that works is to import output css file

**Comment by ShaunEvening** [comment]
@rafalwawrzyk I'm still waiting on a reproduction from someone so that I can investigate. Can you provide one?

**Comment by Flower-dev** [comment]
Hey! I have the same problem with TurboRepo and React when I use this example: https://github.com/vercel/turbo/tree/main/examples/design-system and I install tailwindCSS.

**Comment by rafalwawrzyk** [comment]
If we have structure like this in monorepo: 
```
- apps
  - storybook
  - some-project-1
  - some-project-2
```
 
 storybook is a standalone application
 
```
 - packages
   - ui 
```
  In ui package we got all components with stories
Storybook app see the stories without any problem after configuration, but even if we import directly global.css with tailiwnd directives nothing happens. Only way when it's worked is when we import tailwind output from builded project. 

Also to precise this example, some-project-1 and some-project-2 are next apps, when the storybook is based on react app

If you have questions, feel free :)

**Comment by tibo-glamarche** [comment]
I have found the solution to my problem. Turns out I only needed to add this:
`important: "body"`
to the tailwind.config.js. I am using MaterialUI so maybe the styles were conflicting or something. Before I had:
`important: "#root"` 
because I was targeting the root of my React project. I guess it doesn't work with Storybook

**Comment by valentinpalkovic** [comment, marked as answer]
I've put together an example repository, which is a monorepo, based on turbo and has successfully set up Tailwind with Storybook: https://github.com/valentinpalkovic/example-turbo-repo-tailwind-storybook

The folder structure looks like this:

- apps/docs
  - storybook
- packages
  - ui 

Let's go through the steps I've done to make it work:

## Init Tailwind

I have installed and initialized Tailwind in `packages/ui`:

- `cd ./packages/ui`
- `pnpm install tailwind`
- `npx tailwindcss init`

I have followed the tailwind guide to configure `content`, but be aware to use an absolute path glob instead so that Tailwind doesn't mess up if it gets loaded from Storybook

```diff
+ const { join } = require("node:path");

/** @type {import('tailwindcss').Config} */
module.exports = {
-  content: ["./src/**/*.{js,ts,jsx,tsx}"],
+  content: [join(__dirname, "./src/**/*.{js,ts,jsx,tsx}")], 
  theme: {
    extend: {},
  },
  plugins: [],
}
```

## Setup Tailwind for Storybook

In Storybook Vite-based projects, Vite will automatically detect a postcss.config.js file and will apply all the setup plugins.

I have installed and initialized tailwind and the autoprefixer plugin as described in the [official Tailwind docs for Vite](https://tailwindcss.com/docs/guides/vite#react):

- `cd ./apps/docs`
- `pnpm install tailwindcss postcss autoprefixer`
- `npx tailwindcss init -p`

Now delete the generated tailwind config, because we will refer the one in `packages/ui`. Also, I have renamed `postcss.config.js` to `postcss.config.cjs` because `apps/docs` is declared as an ESModule in its package.json (type: module). The content of `postcss.config.cjs` is the following:

```js
/** @type {import('postcss-load-config').Config} */
const config = {
  plugins: [
    require("autoprefixer"),
    require("tailwindcss")("../../packages/ui/tailwind.config.js"),
  ],
};

module.exports = config;
``` 

As the last step, I've created a `preview.js` file in `apps/docs/.storybook` with the following content:

```js
import "../../../packages/ui/index.css";
```

After this, Storybook starts and correctly applies Tailwind classes.

## Reproduction

1. Checkout the aforementioned repo
2. pnpm install
3. pnpm run dev
4. An underlined paragraph should be displayed


Please let me know if it doesn't work in your scenario or you still have questions.

**Reply by Waseemrajashaik** [reply]
i am using typescript for all files including main.ts, preview.ts, tailwind.config.ts and postcss.config.mjs but my tailwind.config.ts theme extends are not showing up in storybook

**Comment by valentinpalkovic** [comment]
Heads up @kylegach. Do you think [this](https://github.com/storybookjs/storybook/issues/25177#issuecomment-1978568613) is worth documenting somewhere?

**Comment by karrui** [comment]
somehow all the above methods did not work, but the method described below works for me.

If you have a monorepo, use NextJS and exports your stylesheet (e.g. `globals.css`) via a package, Storybook (or tailwind) will not compile /inject the css into the stories.  

You **will need to duplicate the stylesheet into your application instead of importing** for the types to be compiled and accessible by storybook. Tore my hair out for a couple hours before trying this last ditch effect (because I thought I was already importing the tailwind css file!!!) and it miraculously worked.

With the following monorepo structure:
```
apps
  └─ web (nextjs app)
      └─ .storybook
tooling
  └─ tailwind
      └─ styles.css
      └─ package.json (exports `styles.css` as `./styles`)
```

## 🚫  Won't work
```tsx
// apps/web/.storybook/preview.ts
import "@org/tailwind/styles"
// Also won't work
import "../../../relative/path/to/tailwind/styles.css"
```
Importing the tailwind styles from outside of root does not seem to work. Hypothesis: might be due to Storybook's NextJS framework plugin (and its corresponding webpack config) not recognizing the import as a CSS file to process.

## ✅  Will work
Create a styles.css file (instead of importing from another package) in the NextJS app's `.storybook` directory with the tailwind directives, and tailwind will begin to work:

```css
/* apps/web/.storybook/styles.css */
@tailwind base;
@tailwind components;
@tailwind utilities;

/* copy whatever else was exported from the package too */
```

```tsx
// apps/web/.storybook/preview.ts
import "./styles.css"
```

**Reply by acrognale** [reply]
In case anyone else stumbles on this, don't try symlinking it either :(

**Comment by zigang93** [comment]
@valentinpalkovic can you have a look on my config.. it is really hard to debug on storybook,
I have reproduce repo included:
https://github.com/storybookjs/storybook/discussions/28731

**Comment by tomavic** [comment]
Guys, attention please

How to use storybok for angular library project in nx library knowing that it uses webpack as config builder and not Vite.

**Reply by tomavic** [reply]
I get `Error: Cannot apply unknown utility class: font-semibold`

**Reply by severinlandolt** [reply]
Same!

**Comment by cyyyu** [comment]
Encountered the same issue today. 
In my scenario I'm having a monorepo, the storybook exists in the same folder of my UI components. Whether you put them in the same folder or host your storybook outside, the idea behind them should be the same.

```
-- packages/
---- ui/
------ .storybook/
------ src/
-------- button/
---------- button.stories.tsx
------ tailwind.config.ts
------ postcss.config.cjs
------ ...
```

After investigate for a while, I found that the root cause is tailwind doesn't know where are the source files to scan classnames from. In the postcss config I have to override the `content` paths of the tailwind config with the **absolute** paths. 
Below is my full config in `.cjs`. With `.mjs` or `.ts` you need to convert the syntax to ESM style and it should be working fine.

```
const { join } = require("node:path");

// Load the tailwind config to get the theme
const tailwindConfig = require("./tailwind.config.ts");

/** @type {import('postcss-load-config').Config} */
const config = {
  plugins: [
    require("autoprefixer"),
    require("tailwindcss")({
      content: [join(__dirname, "src/**/*.{js,ts,jsx,tsx}")],
      theme: tailwindConfig.default.theme,
      plugins: tailwindConfig.default.plugins,
    }),
  ],
};

module.exports = config;
```

Btw changing the `content` to absolute paths in the tailwind config directly doesn't work for me.
